node/service/src/client.rs

```diff
@@ -0,0 +1,502 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+#[cfg(feature = "with-moonbase-runtime")]
+use moonbase_runtime::{opaque::Block, AccountId, Balance, BlockNumber, Hash, Header, Index};
+#[cfg(feature = "with-moonbeam-runtime")]
+use moonbeam_runtime::{opaque::Block, AccountId, Balance, BlockNumber, Hash, Header, Index};
+#[cfg(feature = "with-moonriver-runtime")]
+use moonriver_runtime::{opaque::Block, AccountId, Balance, BlockNumber, Hash, Header, Index};
+use sc_client_api::{Backend as BackendT, BlockchainEvents, KeyIterator};
+use sp_api::{CallApiAt, NumberFor, ProvideRuntimeApi};
+use sp_blockchain::HeaderBackend;
+use sp_consensus::BlockStatus;
+use sp_runtime::{
+	generic::{BlockId, SignedBlock},
+	traits::{BlakeTwo256, Block as BlockT},
+	Justifications,
+};
+use sp_storage::{ChildInfo, PrefixedStorageKey, StorageData, StorageKey};
+use std::sync::Arc;
+
+/// A set of APIs that polkadot-like runtimes must implement.
+pub trait RuntimeApiCollection:
+	sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+	+ sp_api::ApiExt<Block>
+	+ sp_block_builder::BlockBuilder<Block>
+	+ substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Index>
+	+ pallet_transaction_payment_rpc_runtime_api::TransactionPaymentApi<Block, Balance>
+	+ sp_api::Metadata<Block>
+	+ sp_offchain::OffchainWorkerApi<Block>
+	+ sp_session::SessionKeys<Block>
+where
+	<Self as sp_api::ApiExt<Block>>::StateBackend: sp_api::StateBackend<BlakeTwo256>,
+{
+}
+
+impl<Api> RuntimeApiCollection for Api
+where
+	Api: sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+		+ sp_api::ApiExt<Block>
+		+ sp_block_builder::BlockBuilder<Block>
+		+ substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Index>
+		+ pallet_transaction_payment_rpc_runtime_api::TransactionPaymentApi<Block, Balance>
+		+ sp_api::Metadata<Block>
+		+ sp_offchain::OffchainWorkerApi<Block>
+		+ sp_session::SessionKeys<Block>,
+	<Self as sp_api::ApiExt<Block>>::StateBackend: sp_api::StateBackend<BlakeTwo256>,
+{
+}
+
+#[cfg(feature = "with-moonriver-runtime")]
+pub trait EthereumRuntimeApiCollection {}
+#[cfg(feature = "with-moonriver-runtime")]
+impl<Api> EthereumRuntimeApiCollection for Api {}
+
+#[cfg(not(feature = "with-moonriver-runtime"))]
+pub trait EthereumRuntimeApiCollection:
+	fp_rpc::EthereumRuntimeRPCApi<Block>
+	+ moonbeam_rpc_primitives_debug::DebugRuntimeApi<Block>
+	+ moonbeam_rpc_primitives_txpool::TxPoolRuntimeApi<Block>
+{
+}
+
+#[cfg(not(feature = "with-moonriver-runtime"))]
+impl<Api> EthereumRuntimeApiCollection for Api where
+	Api: fp_rpc::EthereumRuntimeRPCApi<Block>
+		+ moonbeam_rpc_primitives_debug::DebugRuntimeApi<Block>
+		+ moonbeam_rpc_primitives_txpool::TxPoolRuntimeApi<Block>
+{
+}
```

Comment:
What's the idea here? Two traits with the same name but different definitions depending on cargo flags seems like it could be pretty confusing. Maybe I don't get the point yet.